### ISSUE TITLE
Support for interceptors in C#

### ISSUE THREAD

**Issue by Horusiath** [grpc/grpc#11646]
At the present moment it's hard to build any type of generic functionality on top of gRPC C# API. It would be great if we have ability to apply interceptors around generic call definitions (ideally for both client  and server side). From what I've seen such API has been exposed for Java and Go platforms.

Example use case for such functionality (which was cause for us to set up this issue) would be ability to set and receive Zipkin trace headers along with all client/server calls.

**Comment by jtattermusch** [comment]
CC @jtattermusch (I will review the API)

**Comment by rogeralsing** [comment]
Some of the infrastructure is in place in the C# API, you can pass your own `CallInvoker` impl to the client, and apply _some_ interception there for outgoing calls.
I am not sure if you can modify headers though.

**Comment by richardcase** [comment]
Is this being worked on?

**Comment by mehrdada** [comment]
@richardcase interceptors are being worked on in a few languages. Are you interested in contributing a proposal or implementation? If not, it would be great if you let us know what your usage scenarios are so that we keep that in mind.

**Comment by Falco20019** [comment]
I would also be interested. Mostly for logging/tracing purpose right now. I need a central logging instance/service. In most cases, i have a couple of services on the same machine and want a centralized log and can even share a logger instance. But right now I haven't found a got point where to intercept outgoing requests and incoming replies without having to add this to every call manually.

**Comment by peterdeme** [comment]
Server side interception would be great for implementing custom OAuth authentication. Or is there any other way I can achieve this?

**Comment by apolcyn** [comment]
@peterdeme there is an existing API that can be used for custom client credentials. See [this discussion on grpc.io](https://groups.google.com/forum/#!msg/grpc-io/7iKNh3UpuhQ/qn3mWkatAAAJ) for some context.

I might take a look at [this c# test](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.IntegrationTesting/MetadataCredentialsTest.cs) for some example usage. Another example of usage is the [implementation of the GoogleAuthInterceptors](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Auth/GoogleAuthInterceptors.cs), with their usage in [the C# interop tests](https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.IntegrationTesting/InteropClient.cs#L111).

**Comment by peterdeme** [comment]
@apolcyn yeah I saw all of these, but this is only for outgoing calls. I'm talking about server-side authentication. As I see, only SSL auth is available currently and there's no way to implement any custom authentication.

**Comment by mehrdada** [comment]
@peterdeme I'm working on adding support of registering interceptors on a channel and receiving callbacks as RPCs come in. Would that resolve your use case?

**Comment by peterdeme** [comment]
Hi @mehrdada,
I guess implementing a custom authentication in the callback, and throwing an exception inside it would solve the case..

**Comment by plaisted** [comment]
@mehrdada 

It would be great to add support for a middleware pipeline similar to aspcore (see https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware) or expressjs.  That way it would be easy to build plugins and customization on top of the c# gRPC framework.  I see the Go version of the library already has this essentially with their interceptor implementation.

As it is I'm having to re-use a bunch of boilerplate code (JWT token validation, logging, custom exception handling, etc) for each rpc endpoint.

**Comment by mehrdada** [comment]
@plaisted thanks for the feedback. Let me check it out and see how it would fit.

**Comment by plaisted** [comment]
@mehrdada 

I hacked together an example regarding what functionality I was looking for at https://github.com/plaisted/grpc-csharp-middleware

I only implemented it for `UnaryServerCallHandler` for a demo but would be simple to extend the concept to the others as well. With the streaming request or response I'm guessing it'd only make sense the have the middleware execute at the beginning and end of the transmission. Executing for every item in the stream seems like overkill.

Anyway the basic idea is that you can create a middleware pipeline like:
```csharp
var pipeline = new PipelineBuilder();
```

you can then add a middleware to the pipeline (delegate in this example although the POC in the repo allows middleware classes to be used as well):
```csharp
pipeline.Use(async (ctx, next) =>
{
    Console.WriteLine(ctx.Request.ToString()); //executed on the request path
    await next(ctx); //continue to next item in pipeline (main RPC function in this case)
    Console.WriteLine(ctx.Response.ToString()); //executed on the return path
});
```

and configure the server to use pipeline on startup:
```csharp
Server server = new Server
{
    MiddlewarePipeline = pipeline.Build(),
    //rest of server config...
};
```

Then when the RPC is called the middleware is executed and the output is seen on console:
```json
{ "name": "you" }
{ "message": "Hello you" }
```

I also added a simple `Dictionary<string, object>` called `Items` to the `ServerCallContext`.  This allows a middleware to pass objects to the actual RPC function. It would allow scenarios where a plugin authenticates a user and adds the user object to the Items dictionary like:
```csharp
//GRPC request handler
public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
{
    var user = context.Items.GetUser(); //GetUser() is extension method
    if (user?.HasClaim("Admin"))
    {
        //do admin
    } else
    {
        //reject
    }
    return Task.FromResult(new HelloReply { Message = "Hello " + request.Name });
}
```

Let me know what you think of this direction.

**Comment by surjith** [comment]
@mehrdada: When I tried with the existing proto files, it doesn't call the middleware. I guess I'm doing something wrong. Should I regenerate the proto files? How did you generate them?

**Comment by mehrdada** [comment]
@surjith Which implementation are you using? Are you using the implementation by @plaisted above? AFAIK no official functionality is merged yet.

**Comment by plaisted** [comment]
Correct, I didn't even fork the main repo to create the demo. You mentioned you were working on something similar so didn't want to put much effort into it if there was something in the works.

If @mehrdada and team agree with the general idea I can clean it up, add test coverage, and submit a PR and we can iterate from there.

**Comment by mehrdada** [comment]
@plaisted Thanks! I am pretty much done with an implementation that is pretty much in the same spirit as yours, but I also want to cover the client side interceptors and keep both sides consistent with each other. Will post it to seek your feedback on it.

**Comment by surjith** [comment]
@mehrdada @plaisted I was using the implementation by @plaisted  to see how it works to find out if this gets implemented officially whether it will help in what I am trying to do (which is custom token based authorisation) . It looks promising and I hope @mehrdada and team gets what they're working on in.

**Comment by mehrdada** [comment]
I have prepared a PR that provides the core functionality here: https://github.com/grpc/grpc/pull/12613
On the server side, the design supports per-service interceptors, as opposed to global per-server. There is some syntactic sugar to be added to make it easy to implement simple interceptors. cc: @plaisted

**Comment by mehrdada** [comment]
@plaisted I think one way to accomplish that on the client side could be via a satellite object attached as a tag to `CallOption` instead of introducing a `ClientInterceptorContext`. This has the advantage of being compatible with custom `CallInvoker` implementations, since we can't change the `CallInvoker` API.

```
next(..., options.WithTag(satelliteData), ...);
```

My current thought is to add a more user-friendly subclass of `ClientInterceptor` that unifies all calls and raises "events" as things happen so that most common interceptor authors would derive from that instead and not need to override five methods to accomplish the same thing, but the core machinery would remain as is, keeping the most flexibility and without hindering performance by adding hooks everywhere if specific interceptors care about that.

I have reservations about opening up `SubstituteHandlers` visibility, since `IServerCallHandler` itself is internal and the dance that I had to go through to get the concrete generic types back from `IServerCallHandler` is too ugly and too tangled to the current implementation that making it public might cause problems for us if we want to change up things without breaking public APIs down the road.

**Comment by Disturbing** [comment]
Just found this - this is **GREAT** work.

Let me know if I can do anything to help get this PR in ASAP. I'll be testing it :).

**Comment by mehrdada** [comment]
@Disturbing Have you looked at https://github.com/grpc/grpc/pull/12613 and the related gRFC proposal https://github.com/grpc/grpc/pull/12613? It contains the core infrastructure for client and server interceptors. Feedback on that is appreciated.

**Comment by plaisted** [comment]
I've archived the `grpc-csharp-middlewhere` repo to avoid confusion. Busy week hope to dig more into your implementation soon.

**Comment by Disturbing** [comment]
Thanks for archiving it, i didn't realized the PR had all the features.

I'll play with #12613 over the weekend, expect feedback in 72 hours-ish. Thanks for pointing me in the right direction.

**Comment by mehrdada** [comment]
@Disturbing Thanks. Looking forward for your feedback. Please consider #12613 as the core underlying hooks on top of which a higher-level interceptor class can be implemented (we have yet to decide whether that's best fit in Grpc.Core assembly or a third-party library 😕).

**Comment by jtattermusch** [comment]
CC @jtattermusch

**Comment by Disturbing** [comment]
@mehrdada 

Overall this implementation is sound and handles all the use cases that I had to implement around gRPC using DynamixProxying.

**The two main use cases are as follows:**

- Pre-call Server acceptance + context. (A) rejecting calls that do not meet generic security requirements and (b) setting up contexts such as Session Data for the upcoming method.

- Post-call Server - checking if "client's" session data to my specific microservice has changed/been updated and generate an updated token to go along with the response trailers.

- Post-call Client - Easily Add headers and get/create headers via call context in one line. @mehrdada  Kudos on the solution .WithItems in the CallContext +1

---

Overall, this functionality adds a lot of value and with a bit more thorough demonstration / documentation, I believe the learning curve of  grpc-csharp will be greatly improved with this addition.\

This and UDP support are the two greatest features of the grpc communities I desire as a user story coming from an x-game developer and generic microservice builder.

**Comment by vadimi** [comment]
@mehrdada do you think we can expect interceptors in `1.9`? that's probably one the most anticipated features

**Comment by mehrdada** [comment]
@vadimi I'll try my best to get it shipped then in 1.9 or 1.9.1 for C#.

**Comment by vadimi** [comment]
hi @mehrdada, is `1.9.1` still the plan?

**Comment by jtattermusch** [comment]
This definitely shouldn't go into 1.9.1, but should wait for the next minor release 1.10.x . 1.9.0 is already out without interceptor support and patch releases are not the right place to add bigger features like this (+ there would be lot of overhead upmerging bigger changes like this to the master branch and then dealing with potential fixes for issues that are found out later - if any).

**Comment by Spintronic** [comment]
Interceptor support is critical to our enterprise. Any progress on this is much appreciated.

**Comment by Falco20019** [comment]
#12613 got merged, so we will see it in 1.10.x :) So thanks to everyone involved in this, since this is also critical to our enterprise.

**Comment by jtattermusch** [comment]
@893949088  the v1.10.0-pre1 nugets have already been pushed (they have the basic interceptor API that gives you the full power, some convenience code to make things easier is coming soon.) - feel free to test it out.

**Comment by Falco20019** [comment]
Thanks :) Saw it half an hour ago, already started implementing the grpc-opentracing for C#

**Comment by mehrdada** [comment]
Closing. Fixed by #12613.

If you have related feedback or specific concerns please feel free to file new issues or post on the mailing list depending on their nature.
